Assertion failed in opencv resize

##### System information (version)
- Python => Python 3.5.2
- OpenCV => 3.1
- bindings: opencv-python-3.4.1.15
- Operating System / Platform => Ubuntun 16.04


##### Detailed description
I use python bindings, and for large sizes, cv2.resize has a failed assertion.
It appeared that opencv crashed if the product of the x and y output size is greater that 2**31. Probably an int 32 that overflows and turns into a negative number.

The bug existed for input size, and has been fixed a while ago, but the same kind of bugs remains for the output size. Indeed, if I specify a ratio of 0.5 for instance, it does not crash

OpenCV(3.4.1) Error: Assertion failed (dsize.area() > 0) in resize, file /io/opencv/modules/imgproc/src/resize.cpp, line 4050
Traceback (most recent call last):
  File "./testcv2.py", line 6, in <module>
    image = cv2.resize(img, (0,0), fx=1, fy=1, interpolation=cv2.INTER_CUBIC)
cv2.error: OpenCV(3.4.1) /io/opencv/modules/imgproc/src/resize.cpp:4050: error: (-215) dsize.area() > 0 in function resize

##### Steps to reproduce

```
import cv2
import numpy as np
img = np.zeros((46341,46341))
image = cv2.resize(img, (0,0), fx=1, fy=1, interpolation=cv2.INTER_CUBIC)
```

--- a/modules/imgproc/src/resize.cpp
+++ b/modules/imgproc/src/resize.cpp
@@ -3672,7 +3672,7 @@ void resize(int src_type,
 {
     CV_INSTRUMENT_REGION()
 
-    CV_Assert((dst_width * dst_height > 0) || (inv_scale_x > 0 && inv_scale_y > 0));
+    CV_Assert((dst_width > 0 && dst_height > 0) || (inv_scale_x > 0 && inv_scale_y > 0));
     if (inv_scale_x < DBL_EPSILON || inv_scale_y < DBL_EPSILON)
     {
         inv_scale_x = static_cast<double>(dst_width) / src_width;
@@ -3684,7 +3684,7 @@ void resize(int src_type,
     int  depth = CV_MAT_DEPTH(src_type), cn = CV_MAT_CN(src_type);
     Size dsize = Size(saturate_cast<int>(src_width*inv_scale_x),
                         saturate_cast<int>(src_height*inv_scale_y));
-    CV_Assert( dsize.area() > 0 );
+    CV_Assert( !dsize.empty() );
 
     CV_IPP_RUN_FAST(ipp_resize(src_data, src_step, src_width, src_height, dst_data, dst_step, dsize.width, dsize.height, inv_scale_x, inv_scale_y, depth, cn, interpolation))
 
@@ -4041,13 +4041,13 @@ void cv::resize( InputArray _src, OutputArray _dst, Size dsize,
 
     Size ssize = _src.size();
 
-    CV_Assert( ssize.width > 0 && ssize.height > 0 );
-    CV_Assert( dsize.area() > 0 || (inv_scale_x > 0 && inv_scale_y > 0) );
+    CV_Assert( !ssize.empty() );
+    CV_Assert( !dsize.empty() || (inv_scale_x > 0 && inv_scale_y > 0) );
     if( dsize.area() == 0 )
     {
         dsize = Size(saturate_cast<int>(ssize.width*inv_scale_x),
                      saturate_cast<int>(ssize.height*inv_scale_y));
-        CV_Assert( dsize.area() > 0 );
+        CV_Assert( !dsize.empty() );
     }
     else
     {

